BV2-14 (feat) : 회원가입 기능 개발

seob7 opened the pull request:
# BV2-14 (feat) : 회원가입 기능 개발

## 구현 목록
- 회원가입 기능
- 회원가입 중복 검증 AOP 개발
- @PersistenceAdapter Annotation 개발
- 기능 별 Interface 구현 ( 인터페이스명 변동 가능성 있습니다 )

## 미구현 목록
- 예외처리 
    - 추후 구현 예정

## To Reviewer
-

seob7 committed:
BV2-13 (feat) : 사용자 input, output interface 구현

BV2-13 (feat) : 사용자 input, output interface 구현

1. 사용자 관련 UseCase, Port, Command 구현

seob7 committed:
BV2-14 (feat) : 회원가입 기능 개발

BV2-14 (feat) : 회원가입 기능 개발

seob7 committed:
BV2-15 (refactor) : 회원가입 기능 수정

BV2-15 (refactor) : 회원가입 기능 수정

- User DB 에 모든 정보를 저장했지만, Accout DB 로 분할하여 저장하도록 변경
- UserRegisterService 에 Command 객체만 접근 가능하도록 변경

shinywoon commented on user/src/main/java/user/application/port/input/UserInfoUseCase.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface UserInfoUseCase {
+
+    UserInfoCommand findByUserId(String userId);
```
명확한 Method 이름을 사용해주세요
getUserInfoBy(String userId)

shinywoon commented on user/src/main/java/user/application/port/input/UserLoginUseCase.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface UserLoginUseCase {
+
+    TokenCommand login(UserLoginCommand userLoginCommand);
```
로그인 방식에 Token을 이용하는 방식도 있지만 아닌 경우도 존재합니다.
만약 로그인 방식이 바뀐다면 해당  Method는 변경이 필요합니다.
좀 더 고민이 필요한 부분인것 같습니다.

shinywoon commented on user/src/main/java/user/application/port/output/UserLoginPort.java:
```diff
@@ -0,0 +1,10 @@
+
+public interface UserLoginPort {
+
+    User findByUserId(String userId);
```
UserInfoPort 와 UserLoginPort 가 동일한 Method를 사용하는데 분리될 이유가 없는것 같습니다.
Method 는 getUser, getUserAccount 등 명확했으면 좋겠습니다.

shinywoon commented on user/src/main/java/user/application/port/output/UserUpdatePort.java:
```diff
@@ -0,0 +1,9 @@
+
+public interface UserUpdatePort {
+
+    boolean update(User user);
```
update 는 결국 getUser + save 와 동일한 기능을 합니다.
분리될 이유는 없는것 같습니다.

shinywoon reviewed (commented):
저는 `output port` 가 왜 나누어져 있는지 모르겠습니다.
모두 `mongo DB` 라는 하나의 DB와 하나의 document로만 이루어져있는데 port를 나눠서 운영하는 이유가 있을까요?
제가 설명한 문서에는 `port`는 외부에서 접근하는 문의 역할이라고 되어있습니다.
`input port` 는 Service에 따라 어디서 들어올지 모르기 때문에 각각 나누어져 있는것이 맞지만 같은 DB로 나가는 port 하나로 통합되어 있는 것이 맞다고 생각됩니다.

shinywoon reviewed (commented):
.

shinywoon commented on user/src/main/java/user/adapter/input/web/request/UserRegisterRequest.java:
```diff
@@ -0,0 +1,35 @@
+@AllArgsConstructor
+public class UserRegisterRequest {
+
+    private String email;
+
+    private String password;
+
+    private String name;
+
+    private String nickName;
+
+    private String department;
+
+    private LocalDate birth;
+
+    private String address;
+
+    private String detailAddress;
+
+    private String basicAddress;
+
+    private String post;
+
+    private Long imageId;
```
validation 이 필요합니다.

shinywoon commented on user/src/main/java/user/adapter/output/UserPersistenceAdapter.java:
```diff
@@ -0,0 +1,27 @@
+    public User save(User user) {
+        user.setStatus(UserStatus.REGISTERED);
+        user.setRole(UserRole.BASIC_USER);
+        user.setRegisteredAt(LocalDateTime.now());
```
서비스의 책임입니다.

shinywoon commented on user/src/main/java/user/application/service/UserRegisterService.java:
```diff
@@ -0,0 +1,41 @@
+                .password(userRegisterCommand.getPassword())
+                .name(userRegisterCommand.getName())
+                .userId(savedUser.getId())
+            .build());
```
Builder 부분은 따로 구현하여 추상화 레벨을 지켜주세요.

shinywoon reviewed (commented):
.

seob7 committed:
BV2-18 (refactor) : 사용자 input, output interface 수정

BV2-18 (refactor) : 사용자 input, output interface 수정

1. 여러 Port 를 하나의 UserPersistencePort 로 병합하였습니다.

2. 일부 메서드명 수정
